src/main/java/com/google/javascript/gents/GentsCodeGenerator.java

```diff
@@ -55,25 +56,30 @@ protected void add(Node n, Context ctx) {
       default:
         break;
     }
-
-    addNewlines(n);
   }
 
   private static final ImmutableSet<Token> TOKENS_TO_ADD_NEWLINES_BEFORE =
       ImmutableSet.of(
           Token.CLASS, Token.EXPORT, Token.FUNCTION, Token.INTERFACE, Token.MEMBER_FUNCTION_DEF);
 
-  /** Add newlines to the generated source */
-  private void addNewlines(Node n) {
-    Node nextNode = n.getNext();
-    if (nextNode != null) {
-      if (nodeComments.getComment(nextNode) == null // Comments already prepend a newline.
-          && TOKENS_TO_ADD_NEWLINES_BEFORE.contains(nextNode.getToken())) {
-        add("\n");
-      }
+  /** Add newlines to the generated source. */
+  private void maybeAddNewline(Node n) {
+    Node previousNode = n.getPrevious();
+    boolean hasComment = nodeComments.hasComment(n) || nodeComments.hasComment(n.getParent());
+    if (isEmptyAndHasComment(previousNode)
```

Comment:
can you add a helper function so you can move the initialization for hasComment onto one line? something like 

```
boolean hasComment = nodeComments.hasComment(n) || nodeComments.hasComment(n.getParent()) || previousLineHasComment(previousNode)
```